idx/memory/meta_tags.go

```diff
@@ -0,0 +1,236 @@
+	h.Write([]byte(builder.String()))
+	return h.Sum32()
+}
+
+// sortQueries sorts all the queries first by key, then by value, then by
+// operator. The order doesn't matter, it only needs to be consistent
+func (m *metaTagRecord) sortQueries() {
+	sort.Slice(m.queries, func(i, j int) bool {
+		if m.queries[i].key == m.queries[j].key {
+			if m.queries[i].value == m.queries[j].value {
+				return m.queries[i].operator < m.queries[j].operator
+			}
+			return m.queries[i].value < m.queries[j].value
+		}
+		return m.queries[i].key < m.queries[j].key
+	})
+}
+
+// matchesQueries compares another tag record's queries to this
+// one's queries. Returns true if they are equal, otherwise false.
+// It is assumed that all the queries are already sorted
+func (m *metaTagRecord) matchesQueries(other metaTagRecord) bool {
+	if len(m.queries) != len(other.queries) {
+		return false
+	}
+
+	for i, query := range m.queries {
+		if query.key != other.queries[i].key {
+			return false
+		}
+
+		if query.operator != other.queries[i].operator {
+			return false
+		}
+
+		if query.value != other.queries[i].value {
+			return false
+		}
+	}
+
+	return true
+}
+
+// hasMetaTags returns true if the meta tag record has one or more
+// meta tags, otherwise it returns false
+func (m *metaTagRecord) hasMetaTags() bool {
+	return len(m.metaTags) > 0
+}
+
+// upsert inserts or updates a meta tag record according to the given specifications
+// it uses the set of tag query expressions as the identity of the record, if a record with the
+// same identity is already present then its meta tags get updated to the specified ones.
+// If the new record contains no meta tags, then the update is equivalent to a delete.
+// Those are the return values:
+// 1) The id at which the new record got inserted
+// 2) Pointer to the inserted metaTagRecord
+// 3) The id of the record that has been replaced if an update was performed
+// 4) Pointer to the metaTagRecord that has been replaced if an update was performed, otherwise nil
+// 5) Error if an error occurred, otherwise it's nil
+func (m metaTagRecords) upsert(metaTags []string, tagQueryExpressions []string) (uint32, *metaTagRecord, uint32, *metaTagRecord, error) {
+	record, err := metaTagRecordFromStrings(metaTags, tagQueryExpressions)
+	if err != nil {
+		return 0, nil, 0, nil, err
+	}
+
+	record.sortQueries()
+	id := record.hashQueries()
+	var oldRecord *metaTagRecord
+	var oldId uint32
+
+	// loop over existing records, starting from id, trying to find one that has
+	// the exact same queries as the one we're upserting
+	for i := uint32(0); i < collisionAvoidanceWindow; i++ {
+		if existingRecord, ok := m[id+i]; ok {
+			if record.matchesQueries(existingRecord) {
+				oldRecord = &existingRecord
+				oldId = id + i
+				delete(m, oldId)
+				break
+			}
+		}
+	}
+
+	if !record.hasMetaTags() {
+		return 0, &record, oldId, oldRecord, nil
+	}
+
+	// now find the best position to insert the new/updated record, starting from id
```

Comment:
often it will, but not necessarily always.

consider this:
- `record1` gets inserted with hash `1`
- `record2` gets inserted with hash `1`, but ID `1` is already in use, so `record2` it gets ID `2`
- `record1` gets deleted
- `record2` gets updated. At this point it is preferrable to reassign it to ID `1`, because that way if there's a future update to `record2` it will be found faster, without have to first check for ID `1` and then ID `2`.

Reply:
in this example, do record1 and record2 have matching queries?

Reply:
they have different queries. but there is a hash collision, so the hashes of those two different sets of queries are the same. this is an unlikely scenario, but not impossible.

Reply:
i see. makes sense